src/Evolution/EventsAndDenseTriggers/EventsAndDenseTriggers.hpp

```diff
@@ -238,6 +244,12 @@ void EventsAndDenseTriggers::run_events(
   ASSERT(heap_size_ != -1, "Not initialized");
   ASSERT(not events_and_triggers_.empty(),
          "Should not be calling run_events with no triggers");
+  using compute_tags = tmpl::remove_duplicates<tmpl::filter<
+      tmpl::flatten<tmpl::transform<
+          tmpl::at<typename Metavariables::factory_creation::factory_classes,
+                   Event>,
+          get_tags<tmpl::_1>>>,
+      db::is_compute_tag<tmpl::_1>>>;
```

Comment:
This was discussed with @wthrowe and @kidder and the preference was this implementation. You need to template the events on the tags they need anyway, either the compute or simple tags. For example, `ObserveFields` could be templated on `EnergyDensityCompute` instead of `EnergyDensity`. This means that each event lists what compute tags it wants (avoids giant magic collections and inferring info from up the tree). With the change you suggested you need to keep the simple tags that are passed as template parameters to the events in sync with the compute tags that you specify to the actions. I think the implementation here is actually much clearer and easier to maintain, which was the consensus with @wthrowe and @kidder

Reply:
Here's a very specific problem I see with the current implementation, and that would be solved by the idea "the code that retrieves tags out of the box doesn't need to know how the tag got in there":

See https://github.com/sxs-collaboration/spectre/blob/develop/src/Elliptic/Executables/Xcts/SolveXcts.hpp#L258.
The compute item `Xcts::Tags::SpacetimeQuantities` takes a list of possible `gr::Tags` that it is able to compute as template parameters, and returns them in a `Variables`, e.g. the spatial metric, lapse, shift and Hamiltonian constraint. In the current implementation I'd have to pass the `Xcts::Tags::SpacetimeQuantities` compute item to every event, making sure I pass the same instantiation to all to avoid conflicts. Then each event may retrieve a subset of the `gr::Tags`. It would be much cleaner to pass `Xcts::Tags::SpacetimeQuantities` to `RunEventsAndTriggers` to make the subitems available to all events, and then each event may retrieve what it needs.

Reply:
I don't understand what you mean by "same instantiation" here so I can't really comment on that. What you are calling a drawback I'm calling a feature. You _should_ have to specify what the event needs access to. I find it much more difficult to have to figure out why or if something was added if it's added in a completely different location for something like observing. Some parts of the evolution code already having actions like "InitializeConstraints" that add partial derivatives, pointwise L2 norms, and 3+1 items into the DataBox. This makes it _really_ difficult to figure out what you actually need to add to observe something. My view is that it is the events job to ensure what it will need is in the box, which is exactly how the current design works. This is exactly what we do with global cache quantities currently. I'll also note that "every Event" -> "every Event that needs it," which so far is ~2.

Reply:
Hm that sort of dependency resolution is not how we currently do things.. All our actions get<> things out of the DataBox assuming it was put in there by initialization code. Even with your current implementation there are plenty of things the event retrieves but doesn't add to the box itself, like coordinates, mesh, time.

I agree the event should specify what it needs access to, and it does since it's templated on the tags that it retrieves and specifies `argument_tags`. What I'm saying is that the event shouldn't be responsible for how those tags get _into_ the box.

With 'same instantiation' I mean that both the event that wants to observe constraint norms, as well as the event that observes lapse, shift etc in the volume, need to be templated on the same `Xcts::Tags::SpacetimeQuantities` compute tag that adds both the constraint norms _and_ the lapse, shift etc to the observation box. This goes against the idea that the an event only adds what it needs.

I also don't love the proliferation of a magic type alias like `compute_tags_for_observation_box` that is read from some other place in the code. At the very least it needs to be documented in the `Event` class, so when a person implements a new event they know what interface is expected from them.